### ISSUE TITLE
Removal of barbarian ware flax breaks saved games with barbarians

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#358]
Under trunk rev 5677, when I attempted to start a saved game that included the barbarian tribe, I received the following error:

map objects: [/home/user/widelands/src/logic/immovable.cc656] immovable type tribe barbarians does not define immovable type "flax1"

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2010-11-04T14:56:44Z
 * owner: chuckw20
 * assignee: nasenbaer
 * the launchpad url was https://bugs.launchpad.net/bugs/670959

**Comment by bunnybot** [comment]
(by chuckw20)
To test:
Load a saved game generated before trunk rev 5654 that contains the barbarian tribe.

**Comment by bunnybot** [comment]
(by hjd)
Confirmed. Some savegames gave another error (though I reckon it's caused by the same underlying reason):
map objects: [/home/user/widelands/src/logic/worker.cc:2877] loading worker: [/home/user/widelands/src/logic/worker_descr.cc:170] ferner has no program 'harvestflax'

**Comment by bunnybot** [comment]
(by sirver)
Changes in the game files always break save game compatibility... this is to be expected as the ruleset of the game changes. Most people in the community think that we should try to keep savegame compatibility; we did this in the past and always tried to keep 3 builds compatible. The truth is that this has never worked, mainly because we did not keep old save games in the repository for consistent testing.

I personally think that savegame compatibility is not such an important feature to have that it allows to taint our code and conf files [compatibility programs, see training sites of barbarian and empire] as it does. I therefore vote for won't fix and set this to discussion once more. IF we decide that we really want to keep save game compatibility, then I strongly suggest that we start right now with building a repository of savegames inside the bzr repo and install a script that runs automatic tests on those save games. The following options are for vote:

1) reset save game compatibility for build 16. Ignore save game compatibility in the future. Arguments: scenario save games are not comptaible anymore because of trigger/events engine -> lua. Lot's of balance and sanity changes in the conf files that make save game compatibilty ugly. And we can get rid of a lot of code.

2) like 1, but start caring for save game compatiblity in build 17 upwards again. Begin building a save game repository with save games from build 16.

3) try to fix this bugs and keep 3 build's save game compatibility like before. Start building a repository of save games RIGHT NOW with all save games we can find from older builds.

I vote for 1). Please all participate.

**Comment by bunnybot** [comment]
(by chuckw20)
Personally, I would vote for  4) "Will Not Fix".   

I believe the extra baggage and effort to remain backwards compatible for old save games is tying a proverbial millstone around your neck for a feature rarely if ever used by the general community.

As a developer, I would rather recreate my save game scenarios for testing than have to deal with trying to maintain old features that are now undesirable as the game evolves.

In other words, I say "Let go of the past.  Make a clean break of it and allow development to focus on the new issues not the unwanted old ones."

**Comment by bunnybot** [comment]
(by nha)
I am absolutely in favour of (3), i.e. keep compatibility for 3 versions. I have some build15 (maybe even build13, need to check) save games that I can contribute to a repository.

The thing is, save game compatibility isn't that big a deal to us developers, but I've heard that some users like to play extensive long games, and if we don't even have build15->16 compatibility, they'll be justified to be mad at us. Sure such a feature may be used only a few times, but those few times can make the difference between a potential future community member staying with us or leaving and taking their anger against us with them.

**Comment by bunnybot** [comment]
(by qcumber-some)
From my point of view, save game compatibility is an absolute must, however I would argue to relate the three save games backwards. In my eyes, compatibility to the last release is enough. Widelands is not released every two months...

Starting a save game repository is a very good idea, and it should also involve "upgraded save games" (save games from build 14, saved again in build 15). From my experience (not with save games, but with business product's project files), a good amount of problems arise from upgrading.

We could add a test set which loads and saves project files, I guess there is not much to do with widelands to enable this. Of course this should not be run automatically on every compilation :-D

**Comment by bunnybot** [comment]
(by tino79)
I am not much into the widelands save/load code, but what about the following:
- integrate a well defined header into savegames, which contains the version which was used for saving (and perhaps to distinct scenario games from normal games)
- this would allow to show only loadable games (or to visually indicate incompatible savegames) without trying to load the whole savegame first

The next step would be to code an "upgrade" layer into the load code:
- here we should define all changes that have to be applied to the old savegame (version x-1) to load it with version x.
- so if the user loads a b16 savegame with build 19 it takes 3 steps or iterations to upgrade the savegame to be compatible with the current engine. This can perhaps take some time, but it should happen only once, the next time the game will be saved in the current format.

Of course this requires some more work to define all the correct upgrade rules, but we would benefit from having the upgrade rules at exactly one point in the code and no legacy code in engine and/or config files any more.

**Comment by bunnybot** [comment]
(by chuckw20)
Okay, I will admit I may be under a mistaken impression on the extent of savegame usage. (#4)

If it is agreed to support backwards compatibility in some fashion, I would vote for #3 with the provisions offered by 
Jens (#6) - "compatibility to the last release is enough" and 
Tino (#7) - "integrate a well defined header into savegames, which contains the version which was used for saving (and perhaps to distinct scenario games from normal games)" and "code an "upgrade" layer into the load code"

I feel their approaches represent the lowest amount of overburden for the project in the long term.

**Comment by bunnybot** [comment]
(by hjd)
I would like to have save game compatibility. Though, Jens' suggestion to keep compatibility with the previous version rather than three versions may be sufficient. This would reduce the compatibility code to just current and previous version. The idea would be that even if build 18 can't read build16 games, you would be able to load them in build17, store them there, and then load them in build18. I'm guessing most users would be upgrading to each version of Widelands and by the time a new version comes along, their old games are either deleted or stored with the current build. This is just a guess though, and does not take into account things like Ubuntu LTS (long term service) which is released every second year and thus skipping several versions.

An easy starting place for building the source code repository would be to create  savegames with each tribe with at least one of each ware, worker and building (including enhanced ones). Both this bug appeared and when the Empire sawmill worker was changed an error appeared, which this approach would catch. It would naive to think this would be able to catch all problems, but it would be a start.

As I understand it, with the introduction of lua, campaign maps WILL be broken no matter what. So at best we would be looking at partial (non-campaign maps) compatibility? 

Lastly, some thoughts about savegames in general. The main purpose of a savegame is to store the progress I have made while playing. The reason I store this is because I want to continue the game some other time and don't want to loose my progress. Like Nicolai mentioned, if users are not able to continue their games after upgrading to the newest version, they will be disappointed they are not able to continue their games. If we are going to break savegame compatibility, the users should be informed, so at least they know that they won't be able to use their savegames. They will then be able to choose if they want to play the latest version or if they want to wrap up some of their older games first. If they are not aware, I am worried we might loose parts of the user base.

**Comment by bunnybot** [comment]
(by sirver)
I feel that we are not word. I do not care about a game I started 3 months ago, I rather start a new one. So having 3 builds compatibility (or even one build compatibility) seems like unimportant for me. I feel that telling our users to finish their current games before updating is perfectly fine for a game (even some commercial games go that route).

However there is a lot of variation one this topic as we can see here. My suggestion is to make a vote on the homepage together with an explanatory news post and get the input from the community. What do others think of this suggestion?

**Comment by bunnybot** [comment]
(by nha)
Well, the concern is mostly with people who use their distribution's package and let those get upgraded.

In any case, this should be fixed in rev5684 of trunk. If you can still find a failing save game, please submit it and re-open this bug report.

**Comment by bunnybot** [comment]
(by sirver)
Released in build16-rc1

**Comment by bunnybot** [comment]
(by hjd)
Nicolai: When you have time, could you take a look at this tread in the forum where some users has problems loading savegames due to flax http://wl.widelands.org/forum/topic/799/?page=1#post-5930. (Yes, it starts as a bit offtopic in another thread, but there seem to be two people having problems with flax.)
One of the users even link to a savegame, but it did not show up in the list of games I could load after downloading it, so I have not been able to confirm this https://docs.google.com/leaf?id=0B9slGDd1sSwyNjk4NmU2YmMtNThhYy00N2QxLWI3MDYtMTRkNTBkNGFlZTBi&hl=en_US

**Comment by bunnybot** [comment]
(by tino79)
I also cannot understand why widelands does not list that savegame, hjd. Workaround: just unzip the .wfg file in your save directory, then widelands finds it, but of course also cannot load it because of the flax.

Another thing ist that the current dev version also cannot load the compatibility game "build13-test1.wgf", which i think was possible when nha comitted 5684?

**Comment by bunnybot** [comment]
(by nasenbaer)
I'll take a look at it

**Comment by bunnybot** [comment]
(by nasenbaer)
indeed Nicolai fixed quite a lot backward compatibility issues in his commit for revision 5684 - just this one fix was missing.

With current revision all compatibility savegames from our compatibility directory in bzr are loadable and can be continued (even the build13 one, which will officially fall out of backward compatibility support with build17 (3 builds backwards)).

Further I will commit a saved game of build16 in next hours to complete our compatibility directory for future references.

**Comment by bunnybot** [comment]
(by sirver)
Released in build17-rc1.
